TestThreadAuthentication.test_curve_user_id appears to be racy

This is a frustrating issue as it only appears on Ubuntu's CI infrastructure and not at all on my machine but I do have a theory as to what is happening. The symptom is that TestThreadAuthentication.test_curve_user_id fails like this:

```
=================================== FAILURES ===================================
_________________ TestThreadAuthentication.test_curve_user_id __________________

self = <zmq.tests.test_auth.TestThreadAuthentication testMethod=test_curve_user_id>

    @skip_pypy
    def test_curve_user_id(self):
        """threaded auth - CURVE"""
        self.auth.allow('127.0.0.1')
        certs = self.load_certs(self.secret_keys_dir)
        server_public, server_secret, client_public, client_secret = certs
    
        self.auth.configure_curve(domain='*', location=self.public_keys_dir)
        server = self.socket(zmq.PULL)
        server.curve_publickey = server_public
        server.curve_secretkey = server_secret
        server.curve_server = True
        client = self.socket(zmq.PUSH)
        client.curve_publickey = client_public
        client.curve_secretkey = client_secret
        client.curve_serverkey = server_public
>       assert self.can_connect(client, server)
E       AssertionError: assert False
E        +  where False = <bound method TestThreadAuthentication.can_connect of <zmq.tests.test_auth.TestThreadAuthentication testMethod=test_curve_user_id>>(<zmq.sugar.socket.Socket object at 0x7fab2a042850>, <zmq.sugar.socket.Socket object at 0x7fab2a042650>)
E        +    where <bound method TestThreadAuthentication.can_connect of <zmq.tests.test_auth.TestThreadAuthentication testMethod=test_curve_user_id>> = <zmq.tests.test_auth.TestThreadAuthentication testMethod=test_curve_user_id>.can_connect

/usr/lib/python3/dist-packages/zmq/tests/test_auth.py:327: AssertionError
```

(full log here https://objectstorage.prodstack4-5.canonical.com/v1/AUTH_77e2ada1e7a84929a74ba3b87153c0ac/autopkgtest-focal/focal/amd64/p/pyzmq/20191101_193354_9ea5a@/log.gz and many more examples linked from http://autopkgtest.ubuntu.com/packages/p/pyzmq/focal/amd64).

What I *think* is happening is that the can_connect call is happening before the configuration of the authentication method has actually completed. The log output for the above failure is:

```
------------------------------ Captured log call -------------------------------
base.py                     53 DEBUG    Starting
thread.py                   86 DEBUG    auth received API command b'ALLOW'
base.py                     73 DEBUG    Allowing 127.0.0.1
base.py                    207 DEBUG    version: b'1.0', request_id: b'1', domain: '', address: '127.0.0.1', identity: b'', mechanism: b'CURVE'
base.py                    222 DEBUG    PASSED (whitelist) address=127.0.0.1
base.py                    375 DEBUG    ZAP reply code=b'400' text=b'Unknown domain'
thread.py                   86 DEBUG    auth received API command b'CURVE'
base.py                    114 DEBUG    Configure curve: *[/tmp/tmpewmxhtfc/public_keys]
base.py                    207 DEBUG    version: b'1.0', request_id: b'1', domain: '', address: '127.0.0.1', identity: b'', mechanism: b'CURVE'
base.py                    222 DEBUG    PASSED (whitelist) address=127.0.0.1
base.py                    356 DEBUG    ALLOWED (CURVE) domain=* client_key=b'iHzx::[5<yCgiHW&3&St(n3jRJ}pg&6ov.TOH7s^'
base.py                    375 DEBUG    ZAP reply code=b'200' text=b'OK'
thread.py                   86 DEBUG    auth received API command b'TERMINATE'
```

You can see that a connection is rejected with "Unknown domain" before the CURVE api method message is received, and locally in passing tests I don't see that. There is another request later which succeeds so maybe I'm failing to understand how all this works and I'm barking up the wrong tree but in any case, the test is definitely failing...

--- a/zmq/auth/thread.py
+++ b/zmq/auth/thread.py
@@ -51,7 +51,17 @@ class AuthenticationThread(Thread):
                 break  # interrupted
 
             if self.pipe in socks and socks[self.pipe] == zmq.POLLIN:
-                terminate = self._handle_pipe()
+                # Make sure all API requests are processed before
+                # looking at the ZAP socket.
+                while True:
+                    try:
+                        msg = self.pipe.recv_multipart(flags=zmq.NOBLOCK)
+                    except zmq.Again:
+                        break
+
+                    terminate = self._handle_pipe(msg)
+                    if terminate:
+                        break
                 if terminate:
                     break
 
@@ -69,14 +79,11 @@ class AuthenticationThread(Thread):
         if not msg: return
         self.authenticator.handle_zap_message(msg)
 
-    def _handle_pipe(self):
+    def _handle_pipe(self, msg):
         """
         Handle a message from front-end API.
         """
         terminate = False
-
-        # Get the whole message off the pipe in one go
-        msg = self.pipe.recv_multipart()
 
         if msg is None:
             terminate = True

